File: cryptography/sha/SHA.ts
/*
*
 * SHA-256哈希算法 - 密码学哈希函数
 *
 * 问题：生成消息的固定长度哈希值
 *
 * 注意：此实现使用Node.js内置crypto模块
 * 实际应用应使用成熟的加密库
 *
 * 时间复杂度: O(n)
 * 空间复杂度: O(1)
*/

// @ts-ignore - Node.js crypto module
const crypto = require('crypto');

/*
*
 * SHA-256哈希函数（使用Node.js内置crypto模块）
*/
function sha256Hash(message: string): string {
    // @ts-ignore
    return crypto.createHash('sha256').update(message).digest('hex');
}

/*
*
 * 主函数
*/
function main(): void {
    console.log("=== SHA-256哈希算法 ===");

    const message = "Hello, World!";
    console.log(`消息: ${message}`);

    const hash = sha256Hash(message);
    console.log(`SHA-256: ${hash}`);

    console.log("\n注意：此实现使用Node.js内置crypto模块，");
    console.log("实际应用应使用成熟的加密库。");
}

// 运行测试
main();

File: cryptography/sha/sha.go
/*
*
 * SHA-256哈希算法 - 密码学哈希函数
 *
 * 问题：生成消息的固定长度哈希值
 *
 * 核心思想：
 * - 填充消息到512位的倍数
 * - 分块处理
 * - 位操作和逻辑运算
 *
 * 注意：此实现使用Go内置库
 * 实际应用应使用成熟的加密库
 *
 * 时间复杂度: O(n)
 * 空间复杂度: O(1)
*/

package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

/*
*
 * SHA-256哈希函数（使用Go内置库）
*/
func sha256Hash(message string) string {
	hash := sha256.Sum256([]byte(message))
	return hex.EncodeToString(hash[:])
}

/*
*
 * 主函数
*/
func main() {
	fmt.Println("=== SHA-256哈希算法 ===")

	message := "Hello, World!"
	fmt.Printf("消息: %s\n", message)

	hash := sha256Hash(message)
	fmt.Printf("SHA-256: %s\n", hash)

	fmt.Println("\n注意：此实现使用Go内置库，")
	fmt.Println("实际应用应使用成熟的加密库。")
}
